File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/SmartMockServer.java
package com.crescentflare.smartmock;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.crescentflare.smartmock.model.SmartMockHeaders;
import com.crescentflare.smartmock.model.SmartMockResponse;
import com.crescentflare.smartmock.responsegenerator.SmartMockEndPointFinder;
import com.crescentflare.smartmock.responsegenerator.SmartMockResponseFinder;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Smart mock library: the main server interface
 */
public class SmartMockServer
{
    /**
     * Singleton instance
     */

    public static final SmartMockServer instance = new SmartMockServer();


    /**
     * Members
     */

    private List<String> cookieValues = new ArrayList<>();
    private boolean cookiesEnabled = false;


    /**
     * Initialization
     */

    public SmartMockServer()
    {
    }


    /**
     * Serve the response
     */

    public void obtainResponse(Context context, final String method, final String rootPath, final String path, final String body, final SmartMockHeaders headers, final SmartMockServerCallback completion)
    {
        if (Looper.myLooper() != Looper.getMainLooper())
        {
            completion.onObtainMockResponse(obtainResponseSync(context, method, rootPath, path, body, headers));
            return;
        }
        final Handler handler = new Handler();
        final Context appContext = context.getApplicationContext();
        Thread thread = new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                final SmartMockResponse response = obtainResponseSync(appContext, method, rootPath, path, body, headers);
                handler.post(new Runnable()
                {
                    @Override
                    public void run()
                    {
                        if (completion != null)
                        {
                            completion.onObtainMockResponse(response);
                        }
                    }
                });
            }
        });
        thread.start();
    }

    public SmartMockResponse obtainResponseSync(Context context, String method, String rootPath, String path, String body, SmartMockHeaders headers)
    {
        // Safety checks
        if (body == null)
        {
            body = "";
        }
        if (headers == null)
        {
            headers = SmartMockHeaders.create(null);
        }

        // Fetch parameters from path
        Map<String, String> parameters = new HashMap<>();
        int paramMark = path.indexOf('?');
        if (paramMark >= 0)
        {
            String[] parameterStrings = path.substring(paramMark + 1).split("&");
            for (String parameterString : parameterStrings)
            {
                String[] parameterPair = parameterString.split("=");
                if (parameterPair.length > 1)
                {
                    try
                    {
                        parameters.put(URLDecoder.decode(parameterPair[0], "UTF-8"), URLDecoder.decode(parameterPair[1], "UTF-8"));
                    }
                    catch (UnsupportedEncodingException ignored)
                    {
                    }
                }
            }
            path = path.substring(0, paramMark);
        }
        if (!path.startsWith("/"))
        {
            path = "/" + path;
        }

        // Add cookies (if enabled)
        if (cookiesEnabled)
        {
            for (String value : cookieValues)
            {
                headers.addHeader("Cookie", value);
            }
        }

        // Find location and generate response
        String filePath = SmartMockEndPointFinder.findLocation(context, rootPath, path);
        if (filePath == null)
        {
            SmartMockResponse response = new SmartMockResponse();
            response.setCode(404);
            response.setMimeType("text/plain");
            response.setStringBody("Request path not found: " + path + " (within: " + rootPath + ")");
            return response;
        }
        SmartMockResponse response = SmartMockResponseFinder.generateResponse(context, headers, method, path, filePath, parameters, body);
        if (cookiesEnabled)
        {
            String cookieValue = response.getHeaders().getHeaderValue("Set-Cookie");
            if (cookieValue != null && cookieValue.length() > 0)
            {
                applyToCookies(cookieValue);
            }
        }
        return response;
    }


    /**
     * Cookie management
     */

    public void enableCookies(boolean enabled)
    {
        cookiesEnabled = enabled;
    }

    public void clearCookies()
    {
        cookieValues.clear();
    }

    private void applyToCookies(String value)
    {
        String[] splitValues = value.split(";");
        for (String splitValue : splitValues)
        {
            String[] valueSet = splitValue.split("=");
            if (valueSet.length > 0)
            {
                int foundAtIndex = -1;
                for (int i = 0; i < cookieValues.size(); i++)
                {
                    String[] checkValueSet = cookieValues.get(i).split("=");
                    if (checkValueSet.length > 0 && checkValueSet[0].equals(valueSet[0]))
                    {
                        foundAtIndex = i;
                        break;
                    }
                }
                if (foundAtIndex >= 0)
                {
                    cookieValues.set(foundAtIndex, splitValue);
                }
                else
                {
                    cookieValues.add(splitValue);
                }
            }
        }
    }


    /**
     * Listener for asynchronous mock server call
     */

    public interface SmartMockServerCallback
    {
        void onObtainMockResponse(SmartMockResponse response);
    }
}

File: MockLibIOS/SmartMockLib/Classes/SmartMockServer.swift
//
//  SmartMockServer.swift
//  SmartMockServer Pod
//
//  Main library: obtain responses from the mock server endpoints
//

public class SmartMockServer: UIViewController {
    
    // --
    // MARK: Singleton instance
    // --
    
    public static let shared = SmartMockServer()

    
    // --
    // MARK: Members
    // --

    private var cookieValues: [String] = []
    private var cookiesEnabled = false

    
    // --
    // MARK: Serve the response
    // --
    
    public func obtainResponse(method: String, rootPath: String, requestPath: String, requestBody: String?, requestHeaders: SmartMockHeaders?, completion: @escaping (_ response: SmartMockResponse) -> Void) {
        if !Thread.isMainThread {
            completion(obtainResponseSync(method: method, rootPath: rootPath, requestPath: requestPath, requestBody: requestBody, requestHeaders: requestHeaders))
            return
        }
        let priority = DispatchQueue.GlobalQueuePriority.default
        DispatchQueue.global(priority: priority).async {
            let result = self.obtainResponseSync(method: method, rootPath: rootPath, requestPath: requestPath, requestBody: requestBody, requestHeaders: requestHeaders)
            DispatchQueue.main.async {
                completion(result)
            }
        }
    }
    
    public func obtainResponseSync(method: String, rootPath: String, requestPath: String, requestBody: String?, requestHeaders: SmartMockHeaders?) -> SmartMockResponse {
        // Safety checks
        let body = requestBody ?? ""
        let headers = requestHeaders ?? SmartMockHeaders.makeFromHeaders(nil)
        var path = requestPath
        
        // Fetch parameters from path
        var parameters: [String: String] = [:]
        if let paramMark = path.characters.index(of: "?") {
            let parameterStrings = path.substring(from: path.index(paramMark, offsetBy: 1)).characters.split{ $0 == "&"}.map(String.init)
            for parameterString in parameterStrings {
                let parameterPair = parameterString.characters.split{ $0 == "=" }.map(String.init)
                if parameterPair.count > 1 {
                    parameters[SmartMockStringUtility.urlDecode(parameterPair[0])] = SmartMockStringUtility.urlDecode(parameterPair[1])
                }
            }
            path = path.substring(to: paramMark)
        }
        if !path.hasPrefix("/") {
            path = "/" + path
        }
        
        // Add cookies (if enabled)
        if cookiesEnabled {
            for value in cookieValues {
                headers.addHeader(key: "Cookie", value: value)
            }
        }
        
        // Find location and generate response
        if let filePath = SmartMockEndPointFinder.findLocation(atRootPath: rootPath, checkRequestPath: path) {
            let response = SmartMockResponseFinder.generateResponse(headers: headers, requestMethod: method, requestPath: path, filePath: filePath, requestGetParameters: parameters, requestBody: body)
            if cookiesEnabled {
                if let cookieValue = response.headers.getHeaderValue(key: "Set-Cookie") {
                    if cookieValue.characters.count > 0 {
                        applyToCookies(value: cookieValue)
                    }
                }
            }
            return response
        }
        let response = SmartMockResponse()
        response.code = 404
        response.mimeType = "text/plain"
        response.setStringBody("Request path not found: " + path + " (within: " + rootPath + ")")
        return response
    }
    

    // --
    // MARK: Cookie management
    // --
    
    public func enableCookies(_ enabled: Bool) {
        cookiesEnabled = enabled
    }
    
    public func clearCookies() {
        cookieValues.removeAll()
    }
    
    private func applyToCookies(value: String) {
        let splitValues = value.characters.split{ $0 == ";"}.map(String.init)
        for splitValue in splitValues {
            let valueSet = splitValue.characters.split{ $0 == "=" }.map(String.init)
            if valueSet.count > 0 {
                var foundAtIndex = -1
                for i in 0..<cookieValues.count {
                    let checkValueSet = cookieValues[i].characters.split{ $0 == "=" }.map(String.init)
                    if checkValueSet.count > 0 && checkValueSet[0] == valueSet[0] {
                        foundAtIndex = i
                        break
                    }
                }
                if foundAtIndex >= 0  {
                    cookieValues[foundAtIndex] = splitValue
                } else {
                    cookieValues.append(splitValue)
                }
            }
        }
    }

}
